### PR TITLE
nixos/resolvconf: allow different implementations

### LINKED ISSUE #139526: wg-quick doesn't work with systemd-resolved
### Describe the bug
If using systemd-resolved and using DNS in a Wireguard config, wg-quick will fail to add the DNS server because it uses the ``resolvconf`` command to do so. Because systemd-resolved overrides the ``resolvconf`` command with its own implementation and wg-quick calls the openresolv version of ``resolvconf``, applying a Wireguard config fails.

### Steps To Reproduce
1. Enable systemd-resolved with ``services.resolved.enable = true;``
2. Run ``wg-quick up {path to config}`` with a Wireguard config file with a DNS entry

### Expected behavior
wg-quick is supposed to call systemd-resolved's ``resolvconf`` command instead of openresolv's when systemd-resolved is enabled.

### Notify maintainers

@elseym @ericsagnes @mic92 @zx2c4 @globin @ma27 @d-xo

### Metadata
```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.10.66, NixOS, 21.11 (Porcupine)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.4pre20210908_3c56f62`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

### PR DESCRIPTION
**Pull request by ncfavier**
Fixes https://github.com/NixOS/nixpkgs/issues/139526 and possibly https://github.com/NixOS/nixpkgs/pull/118742#issuecomment-816929646 by adding a `networking.resolvconf.package` option that can be either openresolv or systemd depending on the chosen implementation.

Weakly depends on
- https://github.com/NixOS/nixpkgs/pull/163958 and
- https://github.com/NixOS/nixpkgs/pull/166142

for the other wg-quick improvements.

Note: networkd users should probably be encouraged to define their wireguard config in networkd directly, but there's no reason the wg-quick module shouldn't also work for now.

### CODE REVIEW

File: nixos/modules/config/resolvconf.nix [hunk 1]
```diff
@@ -54,6 +54,14 @@ in
         '';
       };
 
+      package = mkOption {
+        type = types.package;
+        internal = true;
```
Comment (SuperSandro2000) [inline nixos/modules/config/resolvconf.nix:59]: Not sure why this is internal.

Comment (ncfavier) [reply nixos/modules/config/resolvconf.nix]: No good reason.

Comment (ncfavier) [reply nixos/modules/config/resolvconf.nix]: Well, I think my intention was that users shouldn't see this in the manual and think they have to set it. Maybe just say that in the description?

Comment (SuperSandro2000) [reply nixos/modules/config/resolvconf.nix]: I don't know. A good description and default value can get you a long way.

Comment (ncfavier) [reply nixos/modules/config/resolvconf.nix]: I extended the description and set `openresolv` as the default.

**Comment by ncfavier** [comment]
Dependencies resolved!

**Comment by ncfavier** [comment]
@ofborg test wireguard

**Comment by ncfavier** [comment]
I've added a commit to switch dhcpcd to `networking.resolvconf.package`, could you test this please?

@ofborg test wireguard networking

**Comment by SuperSandro2000** [comment]
Just tested this locally and this fixes wg-quick when used together with systemd-resolved.

I am going ahead and merging this now.

**Comment by ncfavier** [comment]
Thanks!

**Pull request merged**